Virtualbox appliance generation

Valtis opened the pull request:
Add virtualbox recipe and virtualbox appliance generation as additional image type. Virtualbox appliances are currently generated for intel-corei7-64 and intel-quark images.

This is, at this stage, primarily meant to trigger CI testing and gather feedback. In no particular order:

* Recipes for virtualbox and libidl (which is dependency for virtualbox) are currently placed under meta-ostro/recipes-core. I'm not sure if this is the correct place.
* The shell function that handles the appliance generation is in image-dsk.bbclass. I'm not sure if this is proper place for this either.
* There was discussion on mailing list, that using intel-quark as the 32-bit variant is likely a bad idea, as it is highly machine dependent. Perhaps machine independent 32 bit image should be used instead (intel-corei7-32?)
* libidl recipe was originally in meta-layer, but was removed at some point. Brief googling did not reveal if it was completely removed, or moved to some other layer. The recipe is from here: http://git.yoctoproject.org/cgit/cgit.cgi/poky/tree/meta/packages/libidl/libidl_0.8.12.bb?id=8917b6eb2999f8a894c4864246ed34ef56ff00b9
 
Signed-off-by: Erkka Kääriä <erkka.kaaria@intel.com>

mythi commented on meta-ostro/recipes-core/libidl/libidl_0.8.14.bb:
```diff
@@ -0,0 +1,22 @@
+LICENSE = "LGPL"
+DESCRIPTION = "Library for parsing CORBA IDL files"
```
let's whitelist this from meta-gnome

mythi commented on meta-ostro/recipes-core/virtualbox/virtualbox.bb:
```diff
@@ -0,0 +1,75 @@
+SRC_URI[sha256sum] = "f5a44d33a1db911f445b2eb2d22d9293a96a535cba488b5a235577ef868fa63c"
+
+# Disable everything we do not need (which turns out to be almost everything)
+CONFIGS = " \
```
Use ```EXTRA_OECONF```

mythi commented on meta-ostro/recipes-core/virtualbox/virtualbox.bb:
```diff
@@ -0,0 +1,75 @@
+
+}
+
+BBCLASSEXTEND += "native"
```
I'd explicitly make the recipe virtualbox-native.bb

mythi commented on meta-ostro/classes/image-dsk.bbclass:
```diff
@@ -24,16 +24,104 @@
+create_ova() {
+
+
+  if [ "x${MACHINE}x" = "xintel-corei7-64x" -o "x${MACHINE}x" = "xintel-quarkx" ]; then
```
Let's start with intel-corei7-64 only. The if [] can exit if not intel-corei7-64.

mythi commented on meta-ostro/classes/image-dsk.bbclass:
```diff
@@ -24,16 +24,104 @@
+    ${VIRTUALBOX_EXECUTABLE} export ${VM_NAME} --output ${APPLIANCE_NAME} --ovf20
+
+    # delete old symlink
+    rm -f ${SYMLINK_NAME}
```
the symlink handling should come from image.bbclass, no?

mythi commented on meta-ostro/classes/image-dsk.bbclass:
```diff
@@ -24,16 +24,104 @@
+    VM_NAME="${IMAGE_NAME}"
+    RAM="512"
+    VRAM="12"
+    CPU_CORE_COUNT=2
```
It'd be good to add a comment how the defaults got selected and that they can be modified.

mythi commented on meta-ostro/classes/image-dsk.bbclass:
```diff
@@ -24,16 +24,104 @@
+
+    # erase any temp files from previous builds, as these can cause issues
+    # notably NS_ERROR_FACTORY_NOT_REGISTERED (0x80040154)
+    cd /tmp/
```
Is ```VIRTUALBOX_EXECUTABLE``` using both /tmp and ```WORK_FOLDER```?

mythi commented:
@igor-stoppa can you take a look as well and provide your feedback?

Valtis commented:
Addressed some of the feedback:

* libidl is now whitelisted from meta-gnome (I didn't realize it was present already)
* Renamed virtualbox recipe to vboxmanage-native, to better highlight it that it is not a full virtualbox, but rather a minimal subset to enable the VBoxManage tool
* vboxmanage recipe uses EXTRA_OECONF for configuration parameters
* Changed appliance creation shell function only to create virtualbox appliance for 64 bit intel i7 image, and added some comments to explain why certain values are chosen. 

Symlink creation was not addressed, as I need to look into it a bit more to understand how it works.

As for the work directories, WORK_FOLDER is used to store the temporary VM the vboxmanage executable creates and manipulates before exporting the .ova file. It also seems to store some temporary files under /tmp that can cause issues (quark build at least consistently failed for me until I cleared the temp files)

mythi commented on meta-ostro/recipes-core/vboxmanage/vboxmanage-native_5.0.16.bb:
```diff
@@ -0,0 +1,73 @@
+    "
+
+do_configure() {
+    ./configure ${EXTRA_OECONF}
```
```do_configure()``` should automatically run ./configure with the ```EXTRA_OECONF``` parameters so AFAIK ```do_configure()``` can be removed completely.

Valtis replied:
I tried that; compilation failed.

pohly replied:
On Wed, 2016-03-30 at 02:02 -0700, Erkka Kääriä wrote:
> > +do_configure() {
> > +    ./configure ${EXTRA_OECONF}
> 
> I tried that; compilation failed.

It's worth mentioning in the code why something unusual is needed, to
ensure that others know that it's intentional and also to document the
reason.

Valtis replied:
After reading the doc (http://www.yoctoproject.org/docs/2.1/mega-manual/mega-manual.html#new-recipe-configuring-the-recipe), only autotools or cmake based projects handle do_configure() automatically. As this is neither, we must handle configuration manually. 

I added a comment to clarify this.

Valtis commented:
Yes, it's for retriggering testing. I'll keep the message in mind for the future.

Valtis commented:
retest this please

Valtis commented:
* Changed output name from foo.ova to foo.dsk.ova. The image.bbclass symlink creation now recognizes the image and creates a symlink for it automatically. 
* Due to above, removed manual symlink creation
* VM is now unregistered after exporting it. It seems that the host system Virtualbox installation would occasionally pick up the registered virtual machine and list it (not sure why this seems to be only occasional occurence). Unregistering removes it from the list, which avoids spamming the host Virtualbox VM list.

gvancuts commented on meta-ostro/classes/image-dsk.bbclass:
```diff
@@ -24,16 +24,93 @@
+
+
+  VM_NAME="${IMAGE_NAME}"
+  # 512 should be large enough value to run Ostro (Galileo has 256mb ram for example)
```
Nitpick... could you capitalize 'mb' and 'ram' (i.e. 256MB RAM)

gvancuts commented on meta-ostro/classes/image-dsk.bbclass:
```diff
@@ -24,16 +24,93 @@
+  VRAM="12"
+  # As above with ram, 2 cores should be enough for ostro without crippling the host machine
+  CPU_CORE_COUNT=2
+  # required for succesful boot
```
s/succesful/successful

gvancuts commented on meta-ostro/classes/image-dsk.bbclass:
```diff
@@ -24,16 +24,93 @@
+  RAM="512"
+  # min allowed - as we do not run graphical environtment, it should be enough
+  VRAM="12"
+  # As above with ram, 2 cores should be enough for ostro without crippling the host machine
```
s/ram/RAM

gvancuts commented on meta-ostro/classes/image-dsk.bbclass:
```diff
@@ -24,16 +24,93 @@
+  # 512 should be large enough value to run Ostro (Galileo has 256mb ram for example)
+  # but small enough value that even a modest host machine can afford it
+  RAM="512"
+  # min allowed - as we do not run graphical environtment, it should be enough
```
s/environtment/environment

gvancuts commented on meta-ostro/classes/image-dsk.bbclass:
```diff
@@ -24,16 +24,93 @@
+
+  APPLIANCE_NAME="${VM_NAME}.dsk.ova"
+
+  FOLDER_NAME="virtualbox_vm"
```
Wouldn't it be more consistent with Yocto practices to call this folder virtualbox-vm?

Valtis commented:
* Typo fixes
* Changed the /tmp directory deletion to use find

As for the /tmp directory usage: It seems to be hardcoded (https://www.virtualbox.org/browser/vbox/trunk/src/libs/xpcom18a4/ipc/ipcd/shared/src/ipcConfig.cpp#L61). I could patch this to use some directory inside the Yocto build directory, but then we could face similar situation than we had with Java, where binaries grabbed from sstate fail on other build workers, as the build directory paths are different. This isn't also the only usage of /tmp, as 'grep -rIi "/tmp" | wc -l' in the source folder returns  429.

Valtis commented:
retest this please

Valtis commented:
retest this please

Minnowboard tests seem to have failed for unrelated reasons.

mythi commented on meta-ostro/classes/image-dsk.bbclass:
```diff
@@ -24,16 +24,91 @@
+
+  # erase any temp files from previous builds, as these can cause issues
+  # notably NS_ERROR_FACTORY_NOT_REGISTERED (0x80040154)
+  sudo find /tmp -name ".vbox-root*" -type d -prune -exec rm -rf {} \;
```
sudo isn't right. this can be modified such that you only search for files in /tmp/.vbox-error (I'm assuming you wanted to get of the 'permission denied' when scanning /tmp.)

mythi commented on meta-ostro/classes/image-dsk.bbclass:
```diff
@@ -24,16 +24,91 @@
+  cd ${DEPLOY_DIR_IMAGE}
+
+  # export the image
+  ${VIRTUALBOX_EXECUTABLE} export ${VM_NAME} --output ${APPLIANCE_NAME} --ovf20
```
can you avoid ```cd``` with ```--output ${DEPLOY_DIR_IMAGE}/${APPLIANCE_NAME}```

mythi commented on meta-ostro/classes/image-dsk.bbclass:
```diff
@@ -24,16 +24,91 @@
 
 
-COMPRESSIONTYPES_append = " vdi"
+COMPRESSIONTYPES_append = " vdi ova"
```
there were some changes to this. please rebase and work with the conflicts

Valtis commented:
* Rebased on top of the latest meta-ostro
* Replaced the find exec rm - command with plain rm -rf, as the directory name is hardcoded and we can rely on it.
* --output now uses absolute path

mythi commented:
@Valtis Thanks! Looks +1 to me.

@igor-stoppa @jlaako can you please take a look

igor-stoppa commented:
If now we can have .ova files, why still produce the .vdi?
Btw, couldn't the ova use the .dsk directly?

I wouldn't hold back the patch only for what I mentioned, but I think it should be addressed in a couple of rework patches.

Valtis commented:
@igor-stoppa Virtualbox does not accept .dsk files as virtual hard drive (not recognized), so the .vdi is still required. If the file is not needed otherwise, then maybe the .vdi creation could be merged with .ova creation though.

igor-stoppa commented:
@Valtis are you sure? This seems to say otherwise: https://www.virtualbox.org/manual/ch09.html#rawdisk

The .vdi disk image creation should be left as option, I think.
But the creation of the .ova file should not require explicitly selecting the .vdi file.

Valtis commented:
@igor-stoppa VBoxManage: error: Could not get the storage format of the medium '/<snip>/ostro-image-intel-corei7-64-2016-04-07_23-35-47-build-315.dsk' (VERR_NOT_SUPPORTED)

kad commented:
Virtualbox should be able to work with "raw" images. and our dsk image is just raw image.
Virtualbox might be expecting raw images with some other extension, e.g. "*.img" or something similar.

Valtis commented:
Renaming the image to .img has no effect either. Using the page @igor-stoppa linked allows me to create .vmdk file from the .dsk file, and it boots succesfully. However the absolute path to the .dsk file on my machine is present in the .vdsk file, which makes me doubt it would work on other machines.

igor-stoppa commented:
@Valtis did you try to export the machine as appliance? I would expect this phase takes care of whatever might be installation specific, because appliances are meant to be importable wherever.

Valtis commented:
@igor-stoppa No, no I did not for some reason. Exporting it seems to work fine.

Valtis commented:
The appliance now builds its virtual hard drive from the raw image, so the vdi is no longer a dependency.

gvancuts commented:
IIRC from previous investigations, VirtualBox does not support raw images, they have to be converted to a supported format first (and VBoxManage supports converting from a raw image). Using VDI by default is what makes most sense given it's VirtualBox native format. It's still possible to convert that to anything else later (from the vdi or dsk image). I think it would make sense to stop publishing the vdi image if the ova appliance is functional. I would still keep the option in the code to be able to produce a vdi image too if desired.

gvancuts commented on meta-ostro/classes/image-dsk.bbclass:
```diff
@@ -38,6 +39,74 @@ inherit pythonnative
+
+  
+  RAW_IMAGE="${DEPLOY_DIR_IMAGE}/${IMAGE_NAME}.dsk"
+  VIRTUALBOX_IMAGE="${DEPLOY_DIR_IMAGE}/${IMAGE_NAME}.vmdk"
```
Why use VMDK (VMware native format) instead of VDI (VirtualBox native format) here?

Valtis replied:
I just used the approach described here: https://www.virtualbox.org/manual/ch09.html#rawdisk

Looking at the available VBoxManage commands, converthd with -dstformat VDI might be more appropriate though.

However, if the vdi compression (conversion?) already creates a vdi file, and appliance creation creates a vdi file from the raw image, wouldn't this cause issues? At very least this step would be duplicated.

gvancuts replied:
I'm not following the thinking to be honest. That VirtualBox manual page is really about how to give raw access to the host machine HDD. They seem to use an artefact from VMDK to achieve that.

You always need a disk image to be provided as part of an .ova appliance, creating a vmdk using an obscure feature of VirtualBox seems awkward given we have the ability to properly converting our .dsk into .vdi (and also given that the new steps don't really make us gain anything, we're still converting our .dsk image to something else).

What would make sense in my opinion is that we no longer *publish* .vdi images once .ova are available.

kad replied:
created .vmdk file contains about 500 bytes of meta information about existing dsk file. so, on disk, space consumed only once. In case of conversion to vdi first, it means additional space used.

gvancuts replied:
Doesn't that force us to publish an uncompressed .dsk as part of .ova? If so, that would be much larger than publishing an .ova that includes a .vdi file, wouldn't it?

kad replied:
it might be if zip doesn't handle sparse files correctly. @Valtis  what is current sizes of different artifact ?

Valtis replied:
the produced ova file is ~208M, which is the same it was when I was using the vdi file directly. This is considerably smaller than the .dsk image file, which is ~4 gigabytes, but if the image is mostly zeros, I guess the compression works well in this case.

kad replied:
ova internally is zip file as I recall. How much space it actually uses once user downloads and adds ova to his VirtualBox ?

Valtis replied:
About 584M.

Valtis commented:
retest this please

mythi commented:
This is very good now, IMO.

I'll merge it once if the builds/tests pass.

@Valtis we'd need another PR to update docs/local.conf.sample and drop *.vdi.* images

Valtis commented:
retest this please
Strange Jenkins error message.

I enabled the .ova generation in CI, which uncovered some issues with the VBoxManage recipe - turns out this pull request has never actually been tested properly in CI. There were some dependencies to libraries I had present on my machine. These libraries were not listed in the recipe DEPENDS variable or even whitelisted in some cases, and as such the build failed in CI.

Valtis commented:
Pull request was updated to (hopefully) fix the dependency issues. I can now compile it successfully on Ubuntu 14.04 in VM. Majority of the changes are just patching the config\build files to play nice with Yocto build system, but I had to add a new recipe for makeself-script, as it is a dependency and not provided by any of the existing layers as far as I can see.

Valtis commented:
I updated the pr to also remove the vdi-file generation and sha256sum generation for ova. The former is unnecessary and the latter causes the build to fail.

Valtis commented:
retest this please

No idea why I'm getting errors regarding nodejs runtime.

mythi commented:
@valtis there was a change in meta-ostro that needs an update in meta-iot-web too. That's why you see the failure

Valtis commented:
retest this please

gvancuts commented on meta-ostro/classes/image-dsk.bbclass:
```diff
@@ -38,6 +39,74 @@ inherit pythonnative
+  # create virtual hard drive from the raw .dsk image
+  ${VIRTUALBOX_EXECUTABLE} internalcommands createrawvmdk -filename ${VIRTUALBOX_IMAGE} -rawdisk ${RAW_IMAGE}
+  
+  # attach the .vdi image as hard drive
```
We're now creating a vmdk image instead of vdi, aren't we? If so, can you update the comment to reflect that.

gvancuts commented:
Why does sha256sum generation for ova cause the build to fail? That sounds like something we should fix instead of removing altogether?

kad commented:
@gvancuts seems to be bug in image.bbclass in OE core that executes "compression" commands multiple times for the cases where it's not needed.

gvancuts commented:
Thanks @kad , is that bug being tracked/resolved elsewhere (as part of OE-core)? If so, then I'm OK dropping this for now.

kad commented:
Thanks to @Valtis bug now reported https://bugzilla.yoctoproject.org/show_bug.cgi?id=9482
Even proposed patch attached.

Valtis commented:
I fixed the comment issue. The script now also grants read permission for the output files, as they previously had permissions granted only for the owner. This caused a 403 error when attempting to download the files from the server.

Valtis commented:
retest this please

Soletta failure.

Valtis commented:
retest this please

Soletta related failure again.

mythi commented:
@Valtis please rebase and then I'll merge.

Valtis commented:
Tests failed due to a Galileo kernel panic. As this PR does not touch content on the image, I'd say it's unrelated. The previous, pre-rebase tests also passed just fine.

Nevertheless, retest this please

igor-stoppa commented:
@Valtis If you have still the log from the panic, can you please report it?

Valtis commented:
Sure. I'm also re-testing the image locally to see if I can reproduce it.

Valtis commented:
retest this please

Soletta failure again.

okartau commented:
retest this please

Valtis commented:
I renamed the EXTRA_OECONF variable to VBOX_CONF in the recipe, as the old name was misleading (we are not using the automake class), and more importantly, because the recent no-static-libs change added  --disable-static option into EXTRA_OECONF, which broke the configure step as Virtualbox configure script does not recognize this argument.

Valtis commented:
I just have the worst of luck when it comes to tests. This time Beaglebone flashing failed, as the file system was mounted as read-only mid-flashing. It may indicate that the SD card is failing. 

retest this please

Valtis commented:
I rebased the PR to see that all tests are still passing (previous result was good).

Valtis commented:
retest this please

Galileo failed to flash the image.

Valtis commented:
retest this please

Edison failed this time. Seems to be unrelated.

Valtis commented:
retest this please

Build machines failed to rsync the build results.

igor-stoppa commented:
@Valtis:
​This begs the question: should we hire an exorcist ​ ​for your PR or do also everything else fails so often so randomly?​

Valtis commented:
I think this PR would be a good candidate for most random failures before merge.

gvancuts commented:
+1 on hiring an exorcist ;-)

pohly commented:
Builds are finally good. I'll merge.

Regarding exorcists, I believe @Valtis is currently the most suitable person for the job. Kind of ironic that his PR was indeed one of those most affected by this ghost bugs in the testing ;-}
